### PR TITLE
GUI Testing

### PR DESCRIPTION
**Pull request by dschmidt**
This is still WIP but by now it feels like it's worth a first round of reviewing.

The implementation of the command parser is still fragile (splitting by | is not exactly error-proof)
I'm not completely settled on the commands we want to support: this largely depends on feedback from @SamuAlfageme 

As most of it is hidden inside the `SocketApi` class, it should not affect any other part of the application with only one excception:
We just need to give a few widgets names to be able to find them, are you fine with that?

Also let me know if the code does not speak for itself enough and you need any explanation.

### CODE REVIEW

**Comment by dschmidt** [comment]
Jenkins is unhappy about Qt5 code ... but as we're about to drop Qt5 support, I'd consider this a non issue.

**Comment by ogoffart** [comment]
Maybe a small documentation of what the commands do could be nice.  So from that documetation, @SamuAlfageme  would also be able to tell if that is enough for what he wants to do

**Comment by ogoffart** [comment]
We could merge the last commit "Disable Qt4 CI" already!  that'd be useful for other things

**Comment by ogoffart** [comment]
@SamuAlfageme it would be nice to have your input so we can move forward.

**Comment by SamuAlfageme** [comment]
@ogoffart no worries, I'll post here my comments when I sort them out. So far I haven't had time to play as much as I would like with this and most of it was setting a reproducible test env. and conversations with @dschmidt about what can be tested with this, etc.

First real step will probably be to test all the translations and string labels in the client using this.

**Comment by ckamm** [comment]
@SamuAlfageme Is this one still alive?

**Comment by SamuAlfageme** [comment]
@ckamm yeah; it's waiting for further development and some documentation once @dschmidt and I find some time to sit together. It has been pushed for a while now 😕 

See it in action in the talk I gave in the conference: https://www.youtube.com/watch?v=H3FuMeRIR2c & https://www.youtube.com/watch?v=nxQkM6nxW2s and the proof of concept in https://gitlab.com/alfageme/client-GUI-tests/-/jobs/14114030

File: src/gui/socketapi.cpp [hunk 1]
```diff
@@ -506,6 +511,109 @@ void SocketApi::command_SHARE_MENU_TITLE(const QString &, SocketListener* listen
...
+    if (!widget) {
+        job->reject(QLatin1String("widget not found"));
+        return;
+    }
+
+    auto propertyName = job->arguments()[QLatin1String("property")].toString();
+
+    job->resolve(widget->property(propertyName.toLocal8Bit().constData())
+                     .toString()
+                     .toLocal8Bit()
```
Comment (ogoffart) [inline src/gui/socketapi.cpp]: Should be toUtf8()  here and everywhere.  We use UTF8 in the protocol.

File: src/gui/socketapi.cpp [hunk 2]
```diff
@@ -506,6 +511,109 @@ void SocketApi::command_SHARE_MENU_TITLE(const QString &, SocketListener* listen
...
+{
+    auto &arguments = job->arguments();
+
+    auto widget = findWidget(arguments["objectName"].toString());
+    if (!widget) {
+        job->reject(QLatin1String("widget not found"));
+        return;
+    }
+
+    QMetaObject::invokeMethod(widget, arguments["method"].toString().toLocal8Bit().constData());
```
Comment (ogoffart) [inline src/gui/socketapi.cpp]: toString().toLocal8Bit() ?  If that's already a QByteArray, it should have the right encoding.

**Comment by guruz** [comment]
in 5 days, this PR will be 2 years old

**Comment by michaelstingl** [comment]
@guruz old, but still interesting. After Drone setup, we can continue …

**Comment by dschmidt** [comment]
Yes, this is totally still relevant and we are very close to actually being able to use this.

**Comment by dschmidt** [comment]
Updated and rebased. 

If there's nothing fundamentally wrong with the gui testing commands, I'd vote for merging it now. It's pretty much work in progress, but we already use it for our limited tests and it works.
Anyhow I would like to base further feature PRs on this and not end up with a giant branch. Also this has gone out of date several times already.

File: src/gui/socketapi.cpp [hunk 3]
```diff
@@ -58,10 +66,80 @@
 #include <CoreFoundation/CoreFoundation.h>
 #endif
 
+
 // This is the version that is returned when the client asks for the VERSION.
 // The first number should be changed if there is an incompatible change that breaks old clients.
 // The second number should be changed when there are new features.
 #define MIRALL_SOCKET_API_VERSION "1.1"
+#define DEBUG qDebug() << "SocketApi: "
```
Comment (ogoffart) [inline src/gui/socketapi.cpp]: Don't do that.  Use a debug category.  (As it allow to turn the log on and off with a env variable.

But I think you shoud cleanup most debug output.

File: src/gui/socketapi.cpp [hunk 4]
```diff
@@ -1027,6 +1091,177 @@ void SocketApi::command_GET_MENU_ITEMS(const QString &argument, OCC::SocketListe
...
+void SocketApi::command_ASYNC_SET_WIDGET_PROPERTY(const QSharedPointer<SocketApiJob> &job)
+{
+    auto &arguments = job->arguments();
+    auto widget = findWidget(arguments["objectName"].toString());
+    if (!widget) {
+        job->reject(QLatin1String("widget not found"));
+        return;
+    }
+    widget->setProperty(arguments["property"].toString().toUtf8().constData(),
+                        arguments["value"].toString().toUtf8().constData());
```
Comment (ogoffart) [inline src/gui/socketapi.cpp]: The property name is a `char*`  ,  but the property itself is a QVariant,  and you should keep using QString here, not convert to char*

File: src/gui/socketapi.cpp [hunk 5]
```diff
@@ -1027,6 +1091,177 @@ void SocketApi::command_GET_MENU_ITEMS(const QString &argument, OCC::SocketListe
...
+
+            DEBUG << "VALUE: " << value;
+            break;
+        }
+
+        auto tmpObject = var.value<QObject*>();
+        if(tmpObject) {
+            currentObject = tmpObject;
+        } else {
+            DEBUG << "TODO: object not found, what should happen here now?";
```
Comment (ogoffart) [inline src/gui/socketapi.cpp]: Maybe reply with an error on the socket?

File: src/gui/socketapi.cpp [hunk 6]
```diff
@@ -1027,6 +1091,177 @@ void SocketApi::command_GET_MENU_ITEMS(const QString &argument, OCC::SocketListe
...
+        }
+    }
+
+    job->resolve(value.toUtf8().constData());
```
Comment (ogoffart) [inline src/gui/socketapi.cpp]: Resolve takes a QString.  No need to convert.

File: src/gui/socketapi.cpp [hunk 7]
```diff
@@ -1027,6 +1091,177 @@ void SocketApi::command_GET_MENU_ITEMS(const QString &argument, OCC::SocketListe
...
+
+    QObject* currentObject = widget;
+    QString value;
+    for(int i = 0;i<segments.count(); i++) {
+        auto segment = segments.at(i);
+        auto var = currentObject->property(segment.toUtf8().constData());
+
+        if(var.canConvert<QString>()) {
+            var.convert(QMetaType::QString);
+            value = var.value<QString>();
```
Comment (ogoffart) [inline src/gui/socketapi.cpp]: you know that QDebug prints the content of the QVariant even if it is not a QString...

Anyway, this looks like leftover debug that should be cleaned

Comment (dschmidt) [reply src/gui/socketapi.cpp]: It's not converted to print it, it's converted to pass it to `resolve()`

**Comment by jnweiger** [comment]
Yeah merge!

**Pull request merged**